### PR TITLE
feat(helm/nvme): add more knobs to control timeouts and qdepth

### PR DESCRIPTION
**Pull request by tiagolobocastro**
Add more nvme knobs.

## Description
Allow controlling nvme timeouts and q depth.

## Motivation and Context
In some virtual environments (non-nvme), the backend can be slower and lead into timeouts and degraded replicas.
More back-pressure work is required to ensure we don't attempt rebuilds and further IO on a backend which is "stuck", but this will give us some hooks for testing etc.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.

### CODE REVIEW

**Review by hrudaya21 — APPROVED** [review]

**Review by Abhinandan-Purkait — APPROVED** [review]

File: chart/values.yaml [hunk 1]
```diff
@@ -341,8 +341,15 @@ io_engine:
       ptpl: true
       # NVMF target Command Retry Delay for volume target initiators
       hostCmdRetryDelay:
-        # A command retry delay in seconds. A value of 0 means no delay, host may retry immediately
+        # A command retry delay in milliseconds. A value of 0 means no delay, host may retry immediately
         crdt1: 30
+  nvme:
+    ioTimeout: "110s"
+    adminTimeout: "10s"
+    keepAliveTimeout: "10s"
```
Comment (niladrih) [inline chart/values.yaml:349]: Do these need comments to describe how they work?

File: chart/values.yaml [hunk 2]
```diff
@@ -341,8 +341,15 @@ io_engine:
       ptpl: true
       # NVMF target Command Retry Delay for volume target initiators
       hostCmdRetryDelay:
-        # A command retry delay in seconds. A value of 0 means no delay, host may retry immediately
+        # A command retry delay in milliseconds. A value of 0 means no delay, host may retry immediately
         crdt1: 30
+  nvme:
```
Comment (niladrih) [inline chart/values.yaml:346]: Should this go under io_engine?

**Review by niladrih — APPROVED** [review]

File: chart/templates/mayastor/io/io-engine-daemonset.yaml [hunk 3]
```diff
@@ -64,10 +64,14 @@ spec:
         env:
         - name: RUST_LOG
           value: {{ .Values.io_engine.logLevel }}
-        - name: NVME_QPAIR_CONNECT_ASYNC
-          value: "true"
         - name: NVMF_TCP_MAX_QUEUE_DEPTH
-          value: "32"
+          value: "{{ .Values.io_engine.nvme.tcp.maxQueueDepth }}"
+        - name: NVME_TIMEOUT
+          value: "{{ .Values.io_engine.nvme.ioTimeout }}"
```
Comment (Abhinandan-Purkait) [inline chart/templates/mayastor/io/io-engine-daemonset.yaml:70]: Is this time out different from the one we se on the storage class? Do we not need any changes for these in the csi-node?

Comment (tiagolobocastro) [reply chart/templates/mayastor/io/io-engine-daemonset.yaml:70]: Ah I must have squashed the change away, it's on values.yaml, I'll re-add it

File: chart/values.yaml (hunk 2 above)
Comment (tiagolobocastro) [reply chart/values.yaml:346]: Yes

File: chart/templates/mayastor/io/io-engine-daemonset.yaml [hunk 4]
```diff
@@ -64,10 +64,14 @@ spec:
         env:
         - name: RUST_LOG
           value: {{ .Values.io_engine.logLevel }}
-        - name: NVME_QPAIR_CONNECT_ASYNC
-          value: "true"
```
Comment (avishnu) [inline chart/templates/mayastor/io/io-engine-daemonset.yaml:68]: Is this change intentional?

Comment (tiagolobocastro) [reply chart/templates/mayastor/io/io-engine-daemonset.yaml:68]: Yes, it's always enabled for sometime now

File: chart/templates/mayastor/csi/csi-node-daemonset.yaml [hunk 5]
```diff
@@ -78,7 +78,10 @@ spec:
         - "--csi-socket={{ .Values.csi.node.pluginMounthPath }}/{{ .Values.csi.node.socketPath }}"
         - "--node-name=$(MY_NODE_NAME)"
         - "--grpc-endpoint=$(MY_POD_IP):10199"{{ if .Values.csi.node.nvme.io_timeout }}
-        - "--nvme-core-io-timeout={{ .Values.csi.node.nvme.io_timeout }}"{{ end }}{{ if .Values.csi.node.nvme.ctrl_loss_tmo }}
+        - "--nvme-io-timeout={{ .Values.csi.node.nvme.io_timeout }}"
+        - "--nvme-core-io-timeout={{ .Values.csi.node.nvme.io_timeout }}"{{ else }}
+        - "--nvme-core-io-timeout={{ .Values.io_engine.nvme.ioTimeout }}10s"
```
Comment (avishnu) [inline chart/templates/mayastor/csi/csi-node-daemonset.yaml:83]: Seems duplicated?

File: chart/templates/mayastor/csi/csi-node-daemonset.yaml [hunk 6]
```diff
@@ -78,7 +78,10 @@ spec:
         - "--csi-socket={{ .Values.csi.node.pluginMounthPath }}/{{ .Values.csi.node.socketPath }}"
         - "--node-name=$(MY_NODE_NAME)"
         - "--grpc-endpoint=$(MY_POD_IP):10199"{{ if .Values.csi.node.nvme.io_timeout }}
-        - "--nvme-core-io-timeout={{ .Values.csi.node.nvme.io_timeout }}"{{ end }}{{ if .Values.csi.node.nvme.ctrl_loss_tmo }}
+        - "--nvme-io-timeout={{ .Values.csi.node.nvme.io_timeout }}"
```
Comment (avishnu) [inline chart/templates/mayastor/csi/csi-node-daemonset.yaml:81]: Seems duplicated?

File: chart/values.yaml [hunk 7]
```diff
@@ -341,8 +342,24 @@ io_engine:
...
+    # -- Timeout for IOs
+    # This is exagerated value for local disks but we've observed that in
```
Comment (avishnu) [inline chart/values.yaml:349]: typo in exaggerated?

File: chart/README.md [hunk 8]
```diff
@@ -141,6 +141,8 @@ $ helm install my-release openebs/mayastor
 | io_engine.&ZeroWidthSpace;envcontext | Pass additional arguments to the Environment Abstraction Layer. Example: --set {product}.envcontext=iova-mode=pa | `""` |
 | io_engine.&ZeroWidthSpace;logLevel | Log level for the io-engine service | `"info"` |
 | io_engine.&ZeroWidthSpace;nodeSelector | Node selectors to designate storage nodes for diskpool creation Note that if multi-arch images support 'kubernetes.io/arch: amd64' should be removed. | <pre>{<br>"kubernetes.io/arch":"amd64",<br>"openebs.io/engine":"mayastor"<br>}</pre> |
+| io_engine.&ZeroWidthSpace;nvme.&ZeroWidthSpace;ioTimeout | Timeout for IOs This is exagerated value for local disks but we've observed that in shared virtual environments having a higher timeout value is beneficial. In certain cases, you may have to set this to an even higher value. For example, in Hetzner we've had better results setting it to 300s. Please adjust this according to your hardware and needs. | `"110s"` |
```
Comment (avishnu) [inline chart/README.md:144]: typo in exaggerated?

**Comment by tiagolobocastro** [comment]
bors cancel

File: chart/README.md (hunk 8 above)
Comment (tiagolobocastro) [reply chart/README.md:144]: good catch :+1:

File: chart/templates/mayastor/csi/csi-node-daemonset.yaml (hunk 5 above)
Comment (tiagolobocastro) [reply chart/templates/mayastor/csi/csi-node-daemonset.yaml:83]: It's not, but the order is swapped in the else..
I'll keep it the same order

File: chart/templates/storageclass.yaml [hunk 9]
```diff
@@ -10,6 +10,6 @@ metadata:
 parameters:
   repl: {{ .Values.storageClass.parameters.repl | toString | quote }}
   protocol: 'nvmf'
-  ioTimeout: '60'
+  # ioTimeout: '{{ .Values.csi.node.nvme.io_timeout }}'
```
[the remark this answers is not in the document]
Comment (tiagolobocastro) [reply chart/templates/storageclass.yaml]: It's now removed completely, setting timeout per volume doesn't work well at the moment, so removing it completely.

**Comment by tiagolobocastro** [comment]
bors merge
